### PR TITLE
MTDSA-6831: Service and connector for list state benefits

### PR DESCRIPTION
**Pull request by narendravijayarao**

### CODE REVIEW

File: app/v1/connectors/ListBenefitConnector.scala [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import play.api.http.Status
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.http.HttpClient
+import v1.models.request.listBenefits.ListBenefitsRequest
+import v1.models.response.listBenefits.ListBenefitsResponse
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ListBenefitConnector @Inject()(val http: HttpClient,
```
Comment (BetaDraconis) [inline app/v1/connectors/ListBenefitConnector.scala:30]:
```suggestion
class ListBenefitsConnector @Inject()(val http: HttpClient,
```

File: app/v1/connectors/ListBenefitConnector.scala [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+ * limitations under the License.
+ */
+
+package v1.connectors
+
+import config.AppConfig
+import javax.inject.{Inject, Singleton}
+import play.api.http.Status
+import uk.gov.hmrc.http.HeaderCarrier
+import uk.gov.hmrc.play.bootstrap.http.HttpClient
```
Comment (BetaDraconis) [inline app/v1/connectors/ListBenefitConnector.scala:23]: This is the wrong HttpClient import.

File: app/v1/connectors/ListBenefitConnector.scala (hunk 1 above)
Comment (BetaDraconis) [inline app/v1/connectors/ListBenefitConnector.scala:30]: I think we could maybe just use the DeleteRetrieveConnector + service for this endpoint?

File: test/v1/mocks/connectors/MockListBenefitConnector.scala [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import org.scalamock.handlers.CallHandler
+import org.scalamock.scalatest.MockFactory
+import uk.gov.hmrc.http.HeaderCarrier
+import v1.connectors.{DesOutcome, ListBenefitConnector}
+import v1.models.request.listBenefits.ListBenefitsRequest
+import v1.models.response.listBenefits.ListBenefitsResponse
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait MockListBenefitConnector extends MockFactory {
```
Comment (BetaDraconis) [inline test/v1/mocks/connectors/MockListBenefitConnector.scala:28]:
```suggestion
trait MockListBenefitsConnector extends MockFactory {
```

File: test/v1/mocks/connectors/MockListBenefitConnector.scala [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+  val mockListBenefitConnector: ListBenefitConnector = mock[ListBenefitConnector]
```
Comment (BetaDraconis) [inline test/v1/mocks/connectors/MockListBenefitConnector.scala:30]:
```suggestion
  val mockListBenefitsConnector: ListBenefitsConnector = mock[ListBenefitsConnector]
```

File: test/v1/mocks/connectors/MockListBenefitConnector.scala [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+  object MockListBenefitConnector {
```
Comment (BetaDraconis) [inline test/v1/mocks/connectors/MockListBenefitConnector.scala:32]:
```suggestion
  object MockListBenefitsConnector {
```

File: test/v1/mocks/connectors/MockListBenefitConnector.scala [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    def listBenefits(requestData: ListBenefitsRequest): CallHandler[Future[DesOutcome[ListBenefitsResponse]]] = {
+      (mockListBenefitConnector
```
Comment (BetaDraconis) [inline test/v1/mocks/connectors/MockListBenefitConnector.scala:35]:
```suggestion
      (mockListBenefitsConnector
```

File: app/v1/services/ListBenefitService.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import v1.models.errors._
+import v1.models.outcomes.ResponseWrapper
+import v1.models.request.listBenefits.ListBenefitsRequest
+import v1.models.response.listBenefits.ListBenefitsResponse
+import v1.support.DesResponseMappingSupport
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ListBenefitService @Inject()(connector: ListBenefitConnector) extends DesResponseMappingSupport with Logging {
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitService.scala:35]:
```suggestion
class ListBenefitsService @Inject()(connector: ListBenefitsConnector) extends DesResponseMappingSupport with Logging {
```

File: app/v1/services/ListBenefitService.scala [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+  def listBenefit(request: ListBenefitsRequest)
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitService.scala:37]:
```suggestion
  def listBenefits(request: ListBenefitsRequest)
```

File: app/v1/services/ListBenefitService.scala [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+                     (implicit hc: HeaderCarrier, ec: ExecutionContext, logContext: EndpointLogContext):
+  Future[Either[ErrorWrapper, ResponseWrapper[ListBenefitsResponse]]] = {
+
+    val result = for {
+      desResponseWrapper <- EitherT(connector.listBenefits(request)).leftMap(mapDesErrors(mappingDesToMtdError))
+    } yield desResponseWrapper.map(des => des)
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitService.scala:43]:
```suggestion
    } yield desResponseWrapper
```

File: app/v1/services/ListBenefitService.scala [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+
+    result.value
+  }
+
+  private def mappingDesToMtdError: Map[String, MtdError] = Map(
+    "INVALID_TAXABLE_ENTITY_ID" -> NinoFormatError,
+    "INVALID_TAX_YEAR" -> TaxYearFormatError,
+    "INVALID_BENEFIT_ID" -> BenefitIdFormatError,
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitService.scala:51]: The spec has this as mapping to DownstreamError

File: app/v1/services/ListBenefitService.scala [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+    "INVALID_CORRELATIONID" -> DownstreamError,
+    "NO_DATA_FOUND" -> NotFoundError,
+    "INVALID_DATE_RANGE" -> RuleTaxYearRangeInvalidError,
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitService.scala:54]: The spec has this mapping to RuleTaxYearNotSupportedError

File: app/v1/services/ListBenefitService.scala [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    "TAX_YEAR_NOT_SUPPORTED" -> RuleTaxYearNotEndedError,
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitService.scala:55]: This also maps to RuleTaxYearNotSupportedError in the spec

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+import v1.controllers.EndpointLogContext
+import v1.mocks.connectors.MockListBenefitConnector
+import v1.models.errors._
+import v1.models.outcomes.ResponseWrapper
+import v1.models.request.listBenefits.ListBenefitsRequest
+import v1.models.response.listBenefits.{CustomerAddedStateBenefits, CustomerIncapacityBenefit, IncapacityBenefit, ListBenefitsResponse, StateBenefits}
+
+import scala.concurrent.Future
+
+class ListBenefitServiceSpec extends ServiceSpec {
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:29]:
```suggestion
class ListBenefitsServiceSpec extends ServiceSpec {
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+        endDate = Some("2020-04-01"),
+        amount = Some(2000.00),
+        taxPaid = None,
+        submittedOn = Some("2019-04-04T01:01:01Z")
+      )
+    )
+  )
+
+
+  trait Test extends MockListBenefitConnector {
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:175]:
```suggestion
  trait Test extends MockListBenefitsConnector {
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+    implicit val logContext: EndpointLogContext = EndpointLogContext("c", "ep")
+
+    val service: ListBenefitService = new ListBenefitService(connector = mockListBenefitConnector)
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:178]:
```suggestion
    val service: ListBenefitsService = new ListBenefitsService(connector = mockListBenefitsConnector)
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  "ListBenefitsService" when {
+    "listBenefits" must {
+      "return correct result for a success" in new Test {
+        val outcome = Right(ResponseWrapper(correlationId, validResponse))
+
+
+        MockListBenefitConnector.listBenefits(requestData)
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:187]:
```suggestion
        MockListBenefitsConnector.listBenefits(requestData)
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 17]
```diff
@@ -0,0 +1,220 @@
...
+          .returns(Future.successful(outcome))
+
+        await(service.listBenefit(requestData)) shouldBe outcome
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:190]:
```suggestion
        await(service.listBenefits(requestData)) shouldBe outcome
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 18]
```diff
@@ -0,0 +1,220 @@
...
+      }
+
+      "map errors according to spec" when {
+
+        def serviceError(desErrorCode: String, error: MtdError): Unit =
+          s"a $desErrorCode error is returned from the service" in new Test {
+
+            MockListBenefitConnector.listBenefits(requestData)
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:198]:
```suggestion
            MockListBenefitsConnector.listBenefits(requestData)
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 19]
```diff
@@ -0,0 +1,220 @@
...
+              .returns(Future.successful(Left(ResponseWrapper(correlationId, DesErrors.single(DesErrorCode(desErrorCode))))))
+
+            await(service.listBenefit(requestData)) shouldBe Left(ErrorWrapper(Some(correlationId), error))
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:201]:
```suggestion
            await(service.listBenefits(requestData)) shouldBe Left(ErrorWrapper(Some(correlationId), error))
```

File: test/v1/services/ListBenefitServiceSpec.scala [hunk 20]
```diff
@@ -0,0 +1,220 @@
...
+        val input = Seq(
+          ("INVALID_TAXABLE_ENTITY_ID", NinoFormatError),
+          ("INVALID_TAX_YEAR", TaxYearFormatError),
+          ("INVALID_BENEFIT_ID", BenefitIdFormatError),
+          ("INVALID_CORRELATIONID", DownstreamError),
+          ("NO_DATA_FOUND", NotFoundError),
+          ("INVALID_DATE_RANGE", RuleTaxYearRangeInvalidError),
+          ("TAX_YEAR_NOT_SUPPORTED", RuleTaxYearNotEndedError),
+          ("SERVER_ERROR", DownstreamError),
+          ("SERVICE_UNAVAILABLE", DownstreamError)
```
Comment (BetaDraconis) [inline test/v1/services/ListBenefitServiceSpec.scala:213]: This is built to match the code instead of the technical spec.

**Review by BetaDraconis — CHANGES_REQUESTED** [review]
Naming inconsistencies + invalid service error mapping.

File: app/v1/services/ListBenefitService.scala (hunk 10 above)
Comment (narendravijayarao) [reply app/v1/services/ListBenefitService.scala:51]: What do you mean??

Comment (BetaDraconis) [reply app/v1/services/ListBenefitService.scala:51]: read the spec

Comment (Taiwo-AB) [reply app/v1/services/ListBenefitService.scala:51]: "INVALID_BENEFIT_ID" -> DownstreamError

File: app/v1/services/ListBenefitsService.scala [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+import v1.models.request.listBenefits.ListBenefitsRequest
+import v1.models.response.listBenefits.ListBenefitsResponse
+import v1.support.DesResponseMappingSupport
+
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ListBenefitsService @Inject()(connector: ListBenefitsConnector) extends DesResponseMappingSupport with Logging {
+
+  def listBenefit(request: ListBenefitsRequest)
```
Comment (BetaDraconis) [inline app/v1/services/ListBenefitsService.scala:37]: still says `Benefit`

**Review by BetaDraconis — CHANGES_REQUESTED** [review]

**Pull request merged**